packages/hash/api/src/graphql/resolvers/ontology/property-type.ts

```diff
@@ -32,46 +37,63 @@ export const createPropertyType: ResolverFn<
 };
 
 export const getAllLatestPropertyTypes: ResolverFn<
-  Promise<PersistedPropertyType[]>,
+  Promise<PersistedPropertyTypeWithRefs[]>,
   {},
   LoggedInGraphQLContext,
   {}
-> = async (_, __, { dataSources, user }) => {
+> = async (_, __, { dataSources, user }, info) => {
   const { graphApi } = dataSources;
+  const resolvePropertyTypeReferences =
+    fieldByName(info, "referencedDataTypes") !== undefined ||
+    fieldByName(info, "referencedPropertyTypes") !== undefined;
```

Comment:
I think `referencedDataTypes` and `referencedPropertyTypes` should be resolved in dedicated GQL field resolvers. This prevents us from having to write logic like this everywhere a `PersistedPropertyTypeWithRefs ` is resolvable.

There will be a performance hit for now (we will need to call `PropertyTypeModel.getResolved`/`graphApi.getPropertyTypesByQuery` 3 times), but this should be addressed by caching improved methods in the `graphApi` rather than re-structuring the GQL resolvers in this way.

Reply:
This already is an optimization, we could just pass `255` everytime. This was done [as suggested](https://hashintel.slack.com/archives/C02LG39FJAU/p1662569704909709?thread_ts=1662544273.941539&cid=C02LG39FJAU) (_internal_) by @CiaranMn 
The issue here is, that we need this information in this place to generate a query for the REST API.

Reply:
In the future we probably want to pass the argument to a resolver, but for now we agreed on sticking with this approach.

Reply:
My understanding is that this branching is only required temporarily.

I agree that we should have dedicated field resolvers to avoid this.

Reply:
I don't know why we need the branching and can't implement field resolvers immediately btw

Reply:
I didn't appreciate that this is a temporary GQL resolver, in which case I'm happy for it to stay as it is.

But to be clear I wasn't advocating for an argument to be passed to the REST API or the GQL API. I'm advocating for the `referencedDataTypes` and `referencedPropertyTypes` to be dedicated field resolvers, rather than being returned as part of the same resolver for `PersistedPropertyTypeWithRefs`. I think this is bad practice because everywhere we use this `PersistedPropertyTypeWithRefs` type in the GQL API, we will have to remember to compute `referencedDataTypes` and `referencedPropertyTypes` in the resolver before returning the response, otherwise the GQL API will throw a "cannot return null for a non-nullable field" error.

The alternative is to add dedicated field resolvers for `referencedDataTypes` and `referencedPropertyTypes`. The field resolver for `referencedDataTypes` would look something like:
```TypeScript
export const referencedPropertyTypes: ResolverFn<
  Promise<PersistedPropertyType[]>,
  PersistedPropertyTypeWithRefs,
  LoggedInGraphQLContext,
  {}
> = async ({ propertyTypeVersionedUri }, _, { dataSources }) => {
  const { graphApi } = dataSources;

  const { propertyTypeReferences } = await PropertyTypeModel.getResolved(
    graphApi,
    {
      versionedUri: propertyTypeVersionedUri,
      propertyTypeQueryDepth: 255,
    },
  ).catch((err: AxiosError) => {
    throw new ApolloError(
      `Unable to retrieve property type. ${err.response?.data}`,
      "GET_ERROR",
    );
  });

  return propertyTypeReferences.map(propertyTypeModelToGQL);
};
```

Reply:
Generally, I dislike the latter approach, as `getResolved` is a pretty expensive function, which should be avoided at all costs.

Reply:
a mixed approach is to add dedicated field resolvers which fallback to fetching the `referencedPropertyTypes` if they have not already been resolved (assuming it will be `null` and not an empty array, depends on what happens in other queries that will return a `PersistedPropertyType`)

depends on how we plan to deal with this long-term. we could discuss tomorrow

```ts
export const referencedPropertyTypes: ResolverFn<
  Promise<PersistedPropertyType[]>,
  PersistedPropertyTypeWithRefs,
  LoggedInGraphQLContext,
  {}
> = async (
  { 
    referencedPropertyTypes: alreadyResolvedReferencedTypes, 
    propertyTypeVersionedUri 
  }, 
  _, 
  { dataSources }
) => {
  // the parent resolver already added them, just return them
  if (alreadyResolvedReferencedTypes) {
    return alreadyResolvedReferencedTypes;
  }

  // some query ran without populating this field, we need to go fetch the referenced types
  const { graphApi } = dataSources;

  // etc
};
```

Reply:
But this still has the issue, that you don't know what parameters we want to pass to the REST API, right?

the logic would need to be duplicated for each referenced type (and entity later on). I generalized my approach from above (and made it actually type safe): 2a7dec43e7ab0c9d8934b770094076c3001f5352. This falls back to `0` if the field is not requested and to `255` otherwise.

We need that logic for:
- (maybe later) `DataType`
    - `referencedDataTypes`
- `PropertyType`
    - `referencedDataTypes`
    - `referencedPropertyTypes`
- `EntityType`
    - `referencedDataTypes`
    - `referencedPropertyTypes`
    - `referencedLinkTypes`
    - `referencedEntityTypes`

Not fleshed out, but probably something like
- `Entity`
    - `referencedLinks`
    - `referencedEntities`